src/rez/vendor/version/util.py

```diff
@@ -24,3 +27,126 @@ def dedup(iterable):
     """Removes duplicates from a sorted sequence."""
     for e in groupby(iterable):
         yield e[0]
+
+
+@contextmanager
+def dewildcard(request):
+    deer = WildcardReplacer(request)
+    yield deer
+    deer.clean()
+
+
+class WildcardReplacer(object):
```

Comment:
This really needs docstrings. I am having a lot of trouble determining what this class actually does by looking at the code here. It isn't obvious what on_wildcard/on_version should be, nor when they are called or what they expect to be passed.

Would you not have a class that does this:

```
>>> foo = WildcardReplacer("foo-1.*+<2")
>>> foo.get_unexpanded()
Requirement('foo-1+<2')
>>> foo.get_expanded(Version("1.2.3"))
Requirement('foo-1.2+<2')
```
?

Reply:
I also realised that there are expansions that are no longer possible when we aren't expanding pre-build-env. For example, consider "foo-1+<2.**" - that makes sense with pre-build-env expansion but is completely nonsensical post-build-env expansion. I have an inkling that we're not taking this kind of thing into consideration here? I could be wrong though.